### ISSUE TITLE
Get module ID of Duktape/C call

### ISSUE THREAD

**Issue by fatcerberus** [svaarala/duktape#750]
I would like to implement a system where filenames referenced by a package are relative to the package directory.  In other words if a module says:

```javascript
var img = new Image("./awesome.png");
```

`./awesome.png` is loaded from the package.  In order for this to work, I would need to know the module ID of the module calling the `Image` constructor so I could look up its install path.  There doesn't seem to be a way to do that currently.

**Comment by svaarala** [comment]
One possible implementation would be to associate a module ID or a reference to the module metadata object with each function declared in the module. One could then look up the calling function in the call target (say `Image` constructor) and get access to the module metadata via the calling function. It'd probably be best to provide an API to "get calling module metadata object" which would hide the details of this process.

The downside of this approach is that each function will get an extra property, very often unreferenced, which is awkward for low memory targets (functions are already relatively heavyweight and there's often pressure to trim them even as they're now), so it would need to be an optional feature. Also the nearest caller may not always be the relevant one (consider calling through `Array.prototype.forEach()` for example).

Finally, a function can be exported from a module and there are then two possible behaviors: either resolve filenames relative to the module exporting the symbol, or the call site using the function. Both may be relevant in some cases I'd imagine.

**Comment by svaarala** [comment]
Come to think of it, it might be possible to offset the cost of a new property (say, `.module`) by virtualizing the `.fileName` property of a function. `Function.prototype.fileName` could be a getter which looked up the filename from the module metadata table if the function was loaded from a module.

This approach would be cleanest if every code compilation context was considered a module (even "eval" and program code) so that each function had an associated module metadata object. I'm not sure if this would be possible in practice due to the cost of having a module metadata object for each function (some could be shared, but only if at least the `.fileName` of a function matched an existing metadata object).

**Comment by fatcerberus** [comment]
Note that we have to add some form of module tracking anyway, in order to be compliant with CommonJS (`require` is relative to calling module).

**Comment by svaarala** [comment]
I know - that also is not necessarily very easy to implement for the same reasons :-)

**Comment by svaarala** [comment]
Before making that CommonJS `require()` change I'll need to look up how other implementations handle the corner cases (like exporting `require` from a module, calling it using eval, Array `forEach()`, via a Proxy, etc).

**Comment by fatcerberus** [comment]
As a first test, Node.js behavior seems to to match Duktape's:

pig.js
```javascript
console.log("the pig ate everyone [pig.js]");
require = require("./ape/require");
require("./cow");
```

cow.js
```javascript
console.log("the cow ate the pig [cow.js]");
```

ape/require.js
```javascript
module.exports = require;
```

ape/cow.js
```javascript
console.log("the ape ate the pig [ape/cow.js]");
```

`node` output:
```
C:\temp>node pig.js
the pig ate everyone [pig.js]
the ape ate the pig [ape/cow.js]
```

**Comment by fatcerberus** [comment]
Updated my Node.js installation to 6.1.0 - same behavior.

**Comment by svaarala** [comment]
That's what I would actually expect because implementing the literal call site semantics would need engine capabilities not necessarily always available.

**Comment by fatcerberus** [comment]
In light of this behavior, Node.js's `module.require` is looking pretty redundant. :)

**Comment by svaarala** [comment]
I guess it avoids the need to export the `require` symbol explicitly and ensures it's available for all modules.

**Comment by fatcerberus** [comment]
Anyway, I'm thinking the functionality I'm requesting in this issue is impossible to do with a `require()` implemented purely in Ecmascript.  The only way I can see it working is if I manually augment every function in a module (no real way to automate it) and then walk the callstack as part of path resolution.  Having the module tracking in the engine would probably be much cleaner. :)

I guess the footprint issue is the main sticking point here.

**Comment by svaarala** [comment]
Footprint is always a concern of course, but the exact semantics of "calling module" are not necessarily 100% clear either, although that's more of a technical issue.

One footprint friendly solution would be based on the hidden "currentModule" variable which could be looked up from a function's lexical scope so that a single variable could be visible for all functions declared in the same lexical scope. Looking up the calling module would still need some way of figuring out the relevant function (representing the calling module); after that Duktape would do a slow path variable lookup within the context of that function.

**Comment by fatcerberus** [comment]
One way I thought of this could be done without adding special case code to Duktape (and this allow moving the module code into an extra if that becomes desirable) is if Dujtape/C functions could access the caller's scope chain.

In other words, allow C code to do a direct eval in the context of the calling Ecmascript function.  If there are consecutive C calls, just find the most recent Ecmascript call and use that.

**Comment by svaarala** [comment]
Right, this is in essence what I was also proposing earlier: the module would get a magic variable `__moduleId` (with an internal string key so it can't be accessed by standard code) and there'd be a way to read its contents in the scope of the caller.

**Comment by svaarala** [comment]
This solution would be memory efficient because it wouldn't pollute every function with a moduleId property but would still have the same behavior.

**Comment by fatcerberus** [comment]
It would also avoid `require` having to carry around the module ID with it, which would allow modules to share a single `require()` binding.

**Comment by svaarala** [comment]
Right, though as discussed before, this would change the semantics of an exported `require` function. I think the difference is not that significant in practice.

**Comment by fatcerberus** [comment]
The changed behavior would differ from e.g. Node but it would actually still be CommonJS compliant, so yeah, I don't think there would be an issue. :)

**Comment by svaarala** [comment]
Well, I'm still not sure what the CommonJS requirement actually means (w.r.t. eval, Proxies, etc), as the wording isn't that clear. But in any case, if someone wanted to export the `require` function so that it remembers its original module, it can be done simply as:

```js
exports.require = function (id) { return require(id); };
```

so that the calling function will then be in the exporting module. This should be necessary very, very rarely.

**Comment by fatcerberus** [comment]
Hm, this is probably no longer relevant since there will be no built-in module loader anymore?

**Comment by svaarala** [comment]
But ES6 modules will need application integration so Duktape will need to remain aware of module loading somehow. This may still be useful or even necessary.
